pkg/transport/queuestorage/file.go

```
@@ -1,164 +1,328 @@
+	fileInfo, err = f.fd.Stat()
 	if err != nil {
 		return
 	}
 
-	if fileInfo.Size()+int64(len(data)) > f.maxFileSize {
-		if err := f.Close(); err != nil {
-			log.Err(err).Send()
-		}
-		err = fmt.Errorf("file size limit reached")
+	newSize := fileInfo.Size() + dataSize
+	if newSize > f.maxFileSize {
+		openNew = true
 		return
 	}
 
+	return
+}
+
+// postOps uploads the current file and queues its detail
+func (f *FileWriter) postOps() error {
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("uploading file")
+
+	file, err := os.Open(f.path)
+	if err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to open file for upload")
+		return err
+	}
+
+	if err := f.storage.Upload(f.path, file); err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to upload file")
+		return err
+	}
+
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("queuing file")
+
+	var bb []byte
+	if bb, err = json.Marshal(models.FileUploadMessage{
+		Key:  f.key,
+		Path: f.path,
+	}); err != nil {
+		log.Error().Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to marshal file upload message")
+		return err
+	}
+
+	if err := f.queue.Enqueue(bb); err != nil {
+		log.Error().Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to enqueue file")
+		return err
+	}
+
+	// TODO: Consider removing file after upload
+
+	return nil
+}
+
+// Write writes a line of data to the file. It returns the number
+// of bytes written and an error, if any. Write returns
+// a non-nil error when n != len(b) or a constraint is unmet.
+func (f *FileWriter) Write(data []byte) (n int, err error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.terminated {
+		return 0, fmt.Errorf("file writer is terminated")
+	}
+
+	rowID := ulid.Make().String()
+	if data, err = sjson.SetBytes(data, "__row_id", rowID); err != nil {
+		log.Err(err).Msg("unable to set __row_id in JSON")
```

Should this also return an error instead of just logging it?

---

We want writes to succeed regardless of both attributes. An error from this call should not hinder write.